### ISSUE TITLE
.join within a dynamic task leads to insane "Illegal dynamic reference" message

### ISSUE THREAD

**Issue by pshirshov** [sbt/sbt#4170]
- [X] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/1.x/CONTRIBUTING.md) guidelines

## steps

I'm trying to build a map of ProjectReference -> libraryDependencies.value
So, I come up with the following snippet:

```
    val enumerate1 = {
      Def.taskDyn {
        val refs = loadedBuild.value.allProjectRefs

        val tt = refs.map(_._1).map {
          ref =>
            libraryDependencies.all(ScopeFilter(inProjects(ref)))
              .zipWith(Def.setting(ref)) {case (a,b)=> b -> a}
        }

        tt.join.map(t => t)
      }
    }

//... 
    Def.task {
     println(enumerate1.value)
    }
```

## problem

The code above seems to be correct but it fails to build with a deadly obscure message

```
Illegal dynamic reference: Def
[error]     val enumerate1 = {
[error]         ^
[error] one error found
```

## expectation

Ideally such a thing should work. In case it's impossible to fit such computations into sbt design the error message should be sane at least.

## notes

sbt version: *insert sbt version*

**Comment by dwijnand** [comment]
I wonder if this shares the same route cause as #3110.

**Comment by pshirshov** [comment]
I dunno. Just in case, the right way to solve the problem is to use `.evaluate`: 

```
    val buildDepMap = {
      Def.taskDyn {
        val refs = loadedBuild.value.allProjectRefs

        val tt = refs.map(_._1).map {
          ref =>
            libraryDependencies.all(ScopeFilter(inProjects(ref)))
              .zipWith(Def.setting(ref)) { case (a, b) => b -> a }
        }

        Def.task {
          tt
        }
      }
    }

Def.task {
      val sd = settingsData.in(Global).value
      val depmap = buildDepMap.value.map(_.evaluate(sd)).toMap
      // ...
}
```

So this way you can perform almost arbitrary computations within your tasks without depending on fragile macro magic.

For those who may come here with the same issue, the [SO link](https://stackoverflow.com/a/50436200/5704808)

**Comment by dwijnand** [comment]
Closing on assumption that the fix of #3110 fixed this too.
